samples/kinesis_file_gstreamer_sample_app.cpp

```diff
@@ -0,0 +1,797 @@
+                                                        UINT64 dropped_frame_timecode) {
+    LOG_WARN("Reporting dropped frame. Frame timecode " << dropped_frame_timecode);
+    return STATUS_SUCCESS;
+}
+
+STATUS
+SampleStreamCallbackProvider::fragmentAckReceivedHandler(UINT64 custom_data, STREAM_HANDLE stream_handle,
+                                                         UPLOAD_HANDLE upload_handle, PFragmentAck pFragmentAck) {
+    CustomData *data = reinterpret_cast<CustomData *>(custom_data);
+    if (pFragmentAck->ackType == FRAGMENT_ACK_TYPE_PERSISTED) {
+        std::unique_lock<std::mutex> lk(data->file.mtx);
+        uint64_t last_frag_ts = data->file.last_fragment_ts /
+                                duration_cast<nanoseconds>(milliseconds(DEFAULT_TIMECODE_SCALE_MILLISECONDS)).count();
+        if (last_frag_ts != 0 && last_frag_ts == pFragmentAck->timestamp) {
+            LOG_INFO("Successfully persisted file " << data->file.path);
+        }
+    }
+    LOG_DEBUG("Reporting fragment ack received. Ack timecode " << pFragmentAck->timestamp);
+    return STATUS_SUCCESS;
+}
+
+}  // namespace video
+}  // namespace kinesis
+}  // namespace amazonaws
+}  // namespace com;
+
+static void eos_cb(GstElement *sink, CustomData *data) {
+    if (!data->eos_triggered.load()) {
+        // Media pipeline for one track has ended. Next time eos_cb is called means the entire file has been received.
+        data->eos_triggered = true;
+        data->audio_video_sync_cv.notify_all();
+    } else {
+        {
+            std::unique_lock<std::mutex> lk(data->file.mtx);
+            // store file's last fragment's timestamp.
+            data->file.last_fragment_ts = data->key_frame_pts;
+        }
+        LOG_DEBUG("Terminating pipeline due to EOS");
+        g_main_loop_quit(data->main_loop);
+    }
+}
+
+bool put_frame(shared_ptr<KinesisVideoStream> kinesis_video_stream, void *data, size_t len, UINT64 track_id,
+               const nanoseconds &pts, const nanoseconds &dts, FRAME_FLAGS flags) {
+    Frame frame;
+    frame.flags = flags;
+    frame.decodingTs = static_cast<UINT64>(dts.count()) / DEFAULT_TIME_UNIT_IN_NANOS;
+    frame.presentationTs = static_cast<UINT64>(pts.count()) / DEFAULT_TIME_UNIT_IN_NANOS;
+    frame.duration = 0; // with audio, frame can get as close as 0.01ms
+    frame.size = static_cast<UINT32>(len);
+    frame.frameData = reinterpret_cast<PBYTE>(data);
+    frame.trackId = track_id;
+    return kinesis_video_stream->putFrame(frame);
+}
+
+bool all_stream_started(CustomData *data) {
+    bool started = true;
+    if (data->stream_started.size() < DEFAULT_TOTAL_TRACK_COUNT) {
+        started = false;
+    } else {
+        for (map<int, bool>::iterator it = data->stream_started.begin(); it != data->stream_started.end(); ++it) {
+            if (!it->second) {
+                started = false;
+                break;
+            }
+        }
+    }
+
+    return started;
```

Comment:
in that case, just return true here

Reply:
No, it's alright since it's already protected by the mutex, https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/pull/442/files#diff-2af4741aceb98d75d1fb60eac19ef4f3R316